torch/csrc/Module.cpp

```diff
@@ -537,23 +537,20 @@ PyObject *THPModule_hasDistributed(PyObject *_unused)
 #endif
 }
 
-void destroy_DLPack_PyCapsule(PyObject * obj) {
-  delete (DLTensor*)PyCapsule_GetPointer(obj, "tensor");
-}
-
 PyObject *THPModule_toDLPack(PyObject *_unused, PyObject *data)
 {
   THPUtils_assert(THPModule_isTensor(data), "data must be a Tensor");
   auto atTensor = torch::createTensor(data);
-  DLTensor * dlTensor(new DLTensor);
-  at::toDLPack(atTensor, dlTensor);
-  return PyCapsule_New(dlTensor, "tensor", destroy_DLPack_PyCapsule);
+  DLManagedTensor* dlMTensor = at::toDLPack(atTensor);
+  return PyCapsule_New(dlMTensor, "dltensor", NULL);
 }
 
 PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
 {
-  DLTensor * dlTensor = (DLTensor *)PyCapsule_GetPointer(data, "tensor");
-  at::Tensor atensor = at::fromDLPack(dlTensor);
+  DLManagedTensor * dlMTensor = (DLManagedTensor *)PyCapsule_GetPointer(data, "dltensor");
+  // atensor steals the ownership of the underlying storage
+  at::Tensor atensor = at::fromDLPack(dlMTensor);
+  dlMTensor->destructor(dlMTensor->ctx);
```

Comment:
Deleting dlMTensor now invalidates `atensor` immediately, which will cause uses of it to crash. We eventually need to fix `at::fromDLPack` to delete the Tensor when the aten object is deleted, but for now we can just remove this destructor call.